include/sisl/fds/concurrent_insert_vector.hpp

```diff
@@ -0,0 +1,122 @@
+// user to synchornoize this behavior. This data structure is useful when the user wants to insert data into a vector
+// concurrently in a fast manner and then iterate over the data later. If the user wants a vector implementation which
+// reads concurrently with writer, they can use sisl::ThreadVector. This data structure is provided as a replacement for
+// simplistic cases where insertion and iteration never happen concurrently. As a result it provides better performance
+// than even sisl::ThreadVector and better debuggability.
+//
+// Benchmark shows atleast 10x better performance on more than 4 threads concurrently inserting with mutex.
+//
+template < typename T >
+class ConcurrentInsertVector {
+private:
+    ExitSafeThreadBuffer< std::vector< T >, size_t > tvector_;
+    std::vector< std::vector< T > const* > per_thread_vec_ptrs_;
+
+public:
+    struct iterator {
+        size_t next_thread{0};
+        size_t next_id_in_thread{0};
+        ConcurrentInsertVector const& vec;
+
+        iterator(ConcurrentInsertVector const& v) : vec{v} {}
+        iterator(ConcurrentInsertVector const& v, bool end_iterator) : vec{v} {
+            if (end_iterator) { next_thread = vec.per_thread_vec_ptrs_.size(); }
+        }
+
+        void operator++() {
+            ++next_id_in_thread;
+            if (next_id_in_thread >= vec.per_thread_vec_ptrs_[next_thread]->size()) {
+                ++next_thread;
+                next_id_in_thread = 0;
+            }
+        }
+
+        bool operator==(iterator const& other) const {
+            return (next_thread == other.next_thread) && (next_id_in_thread == other.next_id_in_thread);
+        }
+
+        bool operator!=(iterator const& other) const { return !(*this == other); }
+
+        T const& operator*() const { return vec.per_thread_vec_ptrs_[next_thread]->at(next_id_in_thread); }
+        T const* operator->() const { return &(vec.per_thread_vec_ptrs_[next_thread]->at(next_id_in_thread)); }
+    };
+
+    ConcurrentInsertVector() = default;
+    ConcurrentInsertVector(size_t size) : tvector_{size} {}
+    ConcurrentInsertVector(const ConcurrentInsertVector&) = delete;
+    ConcurrentInsertVector(ConcurrentInsertVector&&) noexcept = delete;
+    ConcurrentInsertVector& operator=(const ConcurrentInsertVector&) = delete;
+    ConcurrentInsertVector& operator=(ConcurrentInsertVector&&) noexcept = delete;
+    ~ConcurrentInsertVector() = default;
+
+    template < typename InputType,
+               typename = typename std::enable_if<
+                   std::is_convertible< typename std::decay< InputType >::type, T >::value >::type >
+    void push_back(InputType&& ele) {
+        tvector_.get()->push_back(std::forward< InputType >(ele));
+    }
+
+    template < class... Args >
+    void emplace_back(Args&&... args) {
+        tvector_.get()->push_back(std::forward< Args >(args)...);
+    }
+
+    iterator begin() {
```

Comment:
nit: Not important. ThreadBuffer can add access_all_threads_with_iter(), where we can add iterator begin and end in ThreadBuffer for each thread related data. This way any list, vector could be supported and we might not need iterator here.

Reply:
ThreadBuffer is a more generic class where it doesn't assume anything about mutation vs reading (unlike ConcurrentInsertVector where write and read are never concurrent). As a result, ThreadBuffer has to take a shared lock to access all thread buffers. If we were to have an iterator in ThreadBuffer it will lose its thread safety. That is why in ConcurrentInsertVector we save the pointer of vector on any read and use that for reading.